### PR TITLE
build(common): Make compatible with moduleResolution: node16

### PR DESCRIPTION
**Pull request by tylerbutler**
Updates the following packages in packages/common to build using tsc-multi, create valid ESM types, and adds an exports field and testing using arethetypeswrong:

- container-definitions
- core-interfaces
- core-utils
- driver-definitions

### CODE REVIEW

File: packages/common/container-definitions/package.json [hunk 1]
```diff
@@ -22,8 +34,9 @@
 		"build:commonjs": "fluid-build . --task commonjs",
 		"build:compile": "fluid-build . --task compile",
 		"build:docs": "fluid-build . --task api",
-		"build:esnext": "tsc --project ./tsconfig.esnext.json",
-		"build:test": "tsc --project ./src/test/types/tsconfig.json",
+		"build:esnext": "tsc-multi --config ../../../common/build/build-common/tsc-multi.esm.json",
+		"build:rename-types": "renamer \"lib/**\" -f .d.ts -r .d.mts --force",
```
Comment (tylerbutler) [inline packages/common/container-definitions/package.json:38]:
```suggestion
```

Comment (tylerbutler) [reply packages/common/container-definitions/package.json:38]: Also need to fix the api-extractor:esnext task

File: fluidBuild.config.cjs [hunk 2]
```diff
@@ -236,6 +236,7 @@ module.exports = {
 				// case.
 				"packages/dds/migration-shim/package.json",
 				"packages/test/functional-tests/package.json",
+				"packages/common/client-utils/package.json",
```
[the remark this answers is not in the document]
Comment (tylerbutler) [reply fluidBuild.config.cjs:235]: Yeah, this was a mistake. Reverted.

File: packages/common/core-utils/package.json [hunk 3]
```diff
@@ -42,7 +54,7 @@
 		"test:coverage": "c8 npm test",
 		"test:mocha": "mocha",
 		"test:mocha:verbose": "cross-env FLUID_TEST_VERBOSE=1 npm run test:mocha",
-		"tsc": "tsc",
+		"tsc": "tsc-multi --config ../../../common/build/build-common/tsc-multi.cjs.json",
 		"typetests:gen": "fluid-type-test-generator",
```
Comment (jason-ha) [inline packages/common/core-utils/package.json:58]: It feels like if we are going to bother with tsc-multi for cjs, then we should also rename the .d.ts files (in another change).

File: packages/common/core-utils/src/test/promiseCache.spec.ts [hunk 4]
```diff
@@ -5,7 +5,7 @@
 
 import { strict as assert } from "node:assert";
 import { SinonFakeTimers, useFakeTimers } from "sinon";
-import { PromiseCache } from "../..";
+import { PromiseCache } from "../promiseCache";
```
Comment (jason-ha) [inline packages/common/core-utils/src/test/promiseCache.spec.ts:9]: Tests that weren't reaching into internal modules should not start now.
Update the config if needed to allow self reference.

**Review by jason-ha — APPROVED** [review]
Recommend using self-reference instead of going internal for test, but okay either way.

**Comment by tylerbutler** [comment]
> Recommend using self-reference instead of going internal for test, but okay either way.

I updated it to import from the package itself, which is self-reference if I understood you correctly. Merging.

**Pull request merged**